[logging] Overhaul dynamo_timed and CompilationMetrics logging.

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139849

Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral committed:
[RFC] [logging] Overhaul dynamo_timed and CompilationMetrics logging.

Summary:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetrixContext during compilation, either directly or via dynamo_timed.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields at the end, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.
And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] [logging] Overhaul dynamo_timed and CompilationMetrics logging."


Summary:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetrixContext during compilation, either directly or via dynamo_timed.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields at the end, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1027,26 +1026,12 @@ def format_guard_failures() -> str:
                 shape_env_guard_count = len(output.shape_env.guards)
                 graph_op_count = output.count_calls()
                 graph_node_count = len(output.graph.nodes)
                 graph_input_count = len(output.placeholders)
```
The changes that follow in this file can be summed up by:
1) Rather than setting fields in a MetricsContext, we pass a dict of fields to the MetricsContext. It's the MetricsContext that will initiate the logging on exit.
2) Several fields are no longer updated here. They're set in the MetricsContext via dyanmo_timed.

masnesral commented on torch/_dynamo/convert_frame.py:
```diff
@@ -697,7 +693,10 @@ def compile_inner(
         with contextlib.ExitStack() as stack:
             stack.enter_context(
                 dynamo_timed(
```
As discussed, new dynamo_timed API

masnesral commented on torch/_dynamo/metrics_context.py:
```diff
@@ -0,0 +1,97 @@
+        """
+        Increment a metric by a given amount.
+        """
+        # TODO: do we want to be safe and grab a lock whenever modifying an entry? For
```
Please opine

ezyang replied:
No, but because I think this context manager should be TLS eventually

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -135,10 +136,8 @@
-frame_phase_timing: Dict[str, Dict[str, float]] = collections.defaultdict(
-    lambda: collections.defaultdict(float)
-)
+# cumulative times for any CompilationMetric field populated by dynamo_timed
```
This is to support `calculate_time_spent()`

ezyang replied:
Questions I will be asking: Does this eventually need to be thread local? Does this get clear()'ed when we exit all dynamo_timed spans?

ezyang replied:
You should suffix this with `_us`

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+    fn_name: Optional[str] = None,
+    log_pt2_compile_event: bool = True,
+    metadata: Optional[Dict[str, object]] = None,
+    # TODO: should this be called compilation_metric instead?
```
This is the name we discussed, but it seems a bit strange since "dynamo_compile" exists outside of OSS...

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+        )
+        if dynamo_compile_column:
+            assert METRICS_CONTEXT is not None
+            # TODO: This conditional is here because we have a case where we try to
```
I guess this is due to the bug discussed this morning. We were previously not logging the remote_fx time to access the autotuning cache. Probably there's still a discussion about what to do to support logging that happens in a subprocess, but perhaps that can be later.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
-            if fwd_only and fail_type is None:
-                _add_time_spent(frame_key, phase_name, time_spent)
-            else:
-                # fwd + bwd compilation stages: inductor_compile, code_gen.
```
All of this logging is moved to the backward compile

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -812,6 +726,11 @@ def to_int_us(v: Optional[float]) -> Optional[int]:
 
+# Version field added to every log. Increment to make it easier to distinguish new
+# vs. old entries when you make a substantive change to how the logs are populated.
+LOG_FORMAT_VERSION = 2
```
Note that the field already exists in the internal thrift definition. We're just not logging it yet. I supposed this should be 1 though.

ezyang replied:
2 seems fine, thanks

masnesral commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                    ), context(), track_graph_compiling(
+                        aot_config, "backward"
+                    ), METRICS_CONTEXT, dynamo_timed(
+                        # TODO: Capture this under "entire_frame_compile" or a new name?
```
Please comment on this TODO. Note that this affects how we accumulate the timings that we report from `print_time_report()`. Previously we'd see something like:

```
TIMING: code_gen:0.82134 inductor_compile:0.92101 backend_compile:3.53894 entire_frame_compile:3.61238 total_wall_time:4.43129
```
With this change, we'd see something like:
```
TIMING: code_gen:0.67811 inductor_compile:0.774 backend_compile:2.73924 entire_frame_compile:3.48979 total_wall_time:3.48979
```
That is, the entire_frame_compile now includes the backward compile. I think that's better, but I wonder if it's actually more confusing.

ezyang replied:
No, you can't put it in `entire_frame_compile`. If you start populating it from this site, we now have no way of distinguishing how much time was spent in Dynamo. To do a full summation, we should use the `duration_us` column, which is ALWAYS populated (well, it's not currently populated in dynamo_timed, but it should be.)

masnesral commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                        log_pt2_compile_event=False,
+                        # TODO: How do we want to record this one in dynamo_compile?
+                        dynamo_compile_column="dynamo_cumulative_compile_time_us",
+                    ):
```
The changes below are analogous to the code in convert_frame: gather up the settings we want to ultimately appear in the CompilationMetrics (not otherwise gathered via dynamo_timed) and put them in the MetricsContext.

masnesral commented on torch/_inductor/remote_cache.py:
```diff
@@ -146,9 +146,8 @@ def __init__(
-            phase_name="remote_fx_graph_cache_get",
-            fwd_only=False,
+            "remote_fx_graph_cache_get",
+            dynamo_compile_column="remote_fx_graph_cache_get_time_us",
```
Do we want log_pt2_compile_event True or False here?

ezyang replied:
Do what is behavior neutral and we can reaudit them separately. BTW, why did you rename these? Is it because periods are now banned from the name here?

masnesral replied:
There's nothing renamed here exactly, I just removed what was previously the function name (`"RemoteFxGraphCache.get"`), because I didn't think it was needed. The only reason I pass a function name is that the previous implementation basically required that you provide both a key (or function name) and a phase_name if you want the phase_name to show up in CompilationMetrics. Now you provide the fn_name as an arg only if you want it. I'm really hoping the new signature is a lot more understandable.

masnesral commented:
@ezyang, this could probably use more testing, but given the time crunch getting feedback sooner than later might speed things up a bit

ezyang commented on torch/fx/experimental/validator.py:
```diff
@@ -534,7 +534,9 @@ def add_assertion(self, e: Union[z3.BoolRef, sympy.Basic]) -> None:
         def validate(self) -> None:
-            with dynamo_timed("TranslationValidator.validate"):
+            with dynamo_timed(
+                "TranslationValidator.validate", log_pt2_compile_event=False
```
This is probably OK to emit a row for, it's O(number of compilations)

masnesral replied:
Can do, it was just False before (because the previous default was False)

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -192,7 +192,7 @@ def read_and_emit(
             return None
 
-        with dynamo_timed(key="TritonBundler.read_and_emit", fwd_only=False):
+        with dynamo_timed("TritonBundler.read_and_emit", log_pt2_compile_event=False):
```
How were you deciding whether or not to `log_pt2_compile_event` here? Because most of these ARE being logged to the event table (https://fburl.com/scuba/pt2_compile_events/f8966fla) and you are now excluding them with these changes.

ezyang replied:
Hmm, but now I see there are some places which had `log_pt2_compile_event=True` so now I am confused if we are logging to the table or not. @jamesjwu can you weigh in? @masnesral I think I am satisfied as long as you demonstrate that the behavior doesn't change. But it does seem like we "accidentally" changed the behavior recently...

ezyang replied:
It kind of feels like you flipped the default here, and you could avoided all the faffing about by... not flipping the default.

masnesral replied:
Yeah because I changed the default of `log_pt2_compile_event` from False to True (because that's what we had in your doc https://docs.google.com/document/d/18ZJADPvGaKU47l7UN_DvN-7wfGyFEWIZlSs9zedaUfk). So now i pass False where it wasn't passed before and just take the default where it was pass as True before. So the intent was nothing changes as far as what's logged.

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1439,7 +1439,9 @@ def load_with_key(
             if (time_saved_ns := compiled_graph._time_taken_ns) is not None:
                 cache_info["time_saved_ns"] = time_saved_ns
-                add_remote_cache_time_saved(time_saved_ns, is_backward)
+                METRICS_CONTEXT.increment(
```
nit: It's somewhat more idiomatic in the prevailing codebase to use a function to get the context, as this makes it easier to do things like TLS correctly (and you will need to eventually make `METRICS_CONTEXT` live in TLS the more thread safe we make PT2).

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -1520,7 +1520,10 @@ def fw_compiler_base(
         ) -> CompiledFxGraph:
-            with dynamo_utils.dynamo_timed("compile_fx.<locals>.fw_compiler_base"):
+            with dynamo_utils.dynamo_timed(
+                "compile_fx.<locals>.fw_compiler_base",
```
welp, apparently periods are not illegal lol :P

ezyang commented on test/dynamo/test_metrics_context.py:
```diff
@@ -0,0 +1,59 @@
+            context.set_once("m2", 2)
+            context.update({"m3": 3, "m4": 4})
+
+        self.assertEqual(self.metrics, {"m1": 1, "m2": 2, "m3": 3, "m4": 4})
```
Yeah, more tests that make it clear what aggregation semantics are would be helpful.

masnesral replied:
I've changed the API to have an explicit overwrite flag (Default to False). Hopefully that makes it more clear. Also new tests below.

ezyang replied:
fwiw you don't actually use overwrite ;)

ezyang replied:
(It's not clear to me when you would ever want to use overwrite)

masnesral replied:
removing

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -915,8 +835,30 @@ def add_compilation_metrics_to_chromium(c: CompilationMetrics):
+        metric = metrics.get(field, None)
+        if not metric:
+            return None
+        return float(metric) / 1e6
```
Instead of converting metric to float first, just do metric / 1e6 which is guaranteed to be float because int / float produces float. Because Python is nuts this will actually happen in higher precision than what you have here.

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
-                entire_frame_compile_time,
-                backend_compile_time,
-                inductor_compile_time,
-                code_gen_time,
```
These four dropped (intentionally)

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
-                dynamo_time_before_restart,
-                guarded_code is not None,
-                possibly_missed_reinplacing_opportunities,
-                remote_cache_time_saved,
```
this one dropped intentionally

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
-                json.dumps(config_dict),
-                True,  # is_forward
-                to_int_ms(remote_fx_graph_cache_get_time),
-                to_int_ms(remote_fx_graph_cache_put_time),
```
these two dropped intentionally

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
+                "dynamo_config": json.dumps(config_dict),
+                "is_forward": True,
+                "start_time_us": start_time_ns // 1000,
+                "duration_us": duration_ns // 1000,
```
Of all the things that should be omitted from manual updates, start time and duration are probably the MOST important to omit. You know what the start/end time are in the decorator, use those values.

masnesral replied:
Agreed, but needs a little refactoring. The exceptions are the same (maybe you comment on those to). If you're ok with it, I'd prefer to do it in a separate PR.

masnesral replied:
added TODO above

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
+                "duration_us": duration_ns // 1000,
+                "triton_compile_time_us": None,  # TODO: instrument
+                "runtime_cudagraphify_time_us": None,  # TODO: instrument in separate event
+                "runtime_triton_autotune_time_us": None,  # TODO: instrument in separate event
```
These need to have TODOs somewhere but they shouldn't be here, you're going to add these by setting aggregation on dynamo_compile at which point you need to NOT be explicitly setting them in the update here (btw, if I have a conflict between update and a dynamo_compile_column, what happens? Hopefully it explodes in a visible way?)

masnesral replied:
I'll put them with the definition of CompilationMetrics. I don't know of anywhere else...

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1131,72 +1100,55 @@ def clean_for_json(d: Dict[str, Any]) -> Dict[str, Any]:
-                ),
-                remote_fx_graph_cache_put_time_us=to_int_us(
-                    remote_fx_graph_cache_put_time
+                "cuda_synchronize_time_us": None,  # TODO: instrument
```
ditto https://github.com/pytorch/pytorch/pull/139849/files?w=1#r1832686620

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -864,7 +863,7 @@ def count_args(code: CodeType) -> int:
     with _use_lazy_graph_module(config.use_lazy_graph_module), compile_context(
         CompileContext(compile_id)
-    ):
+    ), METRICS_CONTEXT:
```
nit: An alternative to grabbing `METRICS_CONTEXT` here is you have the `dynamo_compile` extent relatively close by above, where compile_inner is immediately called afterwards. `dynamo_timed` could generated a return value which then you pass in explicitly as an argument to this function. This would help statically verify that these writes are actually going to a context and not being lost into the ether.

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
-                                aot_config,
+                    ), context(), track_graph_compiling(
+                        aot_config, "backward"
+                    ), METRICS_CONTEXT, dynamo_timed(
```
Why do you have to explicitly enter `METRICS_CONTEXT` here? Shouldn't `dynamo_timed` take care of it?

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                                )
+                        except Exception as e:
+                            fail_type = type(e).__qualname__
+                            fail_reason = str(e)
```
I don't think there's a good reason to force this handling at every call site. `dynamo_timed` is 100% capable of catching exceptions and adding them to the field before continuing propagation.

masnesral replied:
Planning to do in a follow-up diff.

masnesral replied:
added TODO

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                            metrics = {
+                                "compile_id": str(
+                                    torch._guards.CompileContext.current_compile_id()
+                                ),
```
I think `dynamo_timed` should take care of reading compile id from tracing context

masnesral replied:
I'll do it with the same follow-up that handles the start/end times and the exception info.

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                                "start_time_us": start_ns // 1000,
+                                "duration_us": (time.time_ns() - start_ns) // 1000,
+                                "structured_logging_overhead_us": to_int_us(
+                                    torch._logging.get_structured_logging_overhead(),
```
`dynamo_timed` should take care of this

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -182,25 +181,14 @@ def increment_op_count(cnt: int) -> None:
-    if total_by_key:
-        total_by_key["total_wall_time"] = total_wall_time
+    for phase, timing in cumulative_time_spent.items():
+        total_by_key[phase] = float(timing) / 1e9
```
nit: you will get more precision in the final count if you say `timing / 1e9` which will automatically convert to float. By casting first you immediately lose precision.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -182,25 +181,14 @@ def increment_op_count(cnt: int) -> None:
 
 
-# Calculate total time spent so far for each phase
+# Get total time spent so far for each phase
```
While the patient is open, nice specifying that these are SECONDS

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
-    time_saved = time_saved_ns / 1e9
-    _add_time_spent(key, "remote_cache_time_saved", time_saved)
+# Use the following singleton to capture and log CompilationMetrics. Entering the context
+# manager allocates a new record to be logged when it exits. (You should not need to use
```
claim: instead, `dynamo_timed` should be preferentially used to cause new records to be allocated, no reason to use this singleton directly.

masnesral replied:
I think the difficulty for me is there's currently a sensible spot where we set all the fields that should be set "globally" once for the log entry. IMO, it's not as clear if dynamo_timed is quietly creating the context for you to store those in.

ezyang replied:
I think in my ideal terminal state, most of the fields are being set (1) in `dynamo_timed`, so they obviously are always set, or (2) done via aggregation, so they're automatically picked up. OK to leave as is for now.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+#   METRICS_CONTEXT.set("metric_name", value)
+#
+# To set multiple fields:
+#   METRICS_CONTEXT.update({"name1": val1, "name2": val2})
```
Need to specify what happens when metric name is specified multiple times

masnesral replied:
I'll make an `overwrite` bool param defaulted to false.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
-    fwd_only: bool = True,
+    event_name: str,
+    # TODO: do we really want event and fn name?
+    fn_name: Optional[str] = None,
```
No

ezyang replied:
Like, I accept you may need to keep it for BC / keeping tests happy, but in the terminal state I don't think it adds value.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+    """
+    # We're standardizing on microseconds for dynamo_compile timings.
+    if dynamo_compile_column is not None:
+        assert dynamo_compile_column.endswith("_us")
```
thanks, appreciated

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+    metadata: Optional[Dict[str, object]] = None,
+    # TODO: should this be called compilation_metric instead?
+    dynamo_compile_column: Optional[str] = None,
 ):
```
nit: while the patient is open, make this strongly typed (need a return type here, among other things)

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
-    time_spent = float("-inf")
+    mkey = fn_name if fn_name else event_name
+    if mkey not in compilation_time_metrics:
+        compilation_time_metrics[mkey] = []
```
nit: `compilation_time_metrics.setdefault(mkey, [])`

masnesral replied:
this thing is existing

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
-                            )
-                            record_compilation_metrics(metrics)
+        time_spent = end_ns - start_ns
+        compilation_time_metrics[mkey].append(float(time_spent) / 1e9)
```
ditto here, don't explicitly cast to float

ezyang replied:
If it's NOT legacy, then you should be internally storing these as `us`, because that is our standard resolution.

masnesral replied:
Well I don't know if it's "legacy" per se. It's existing functionality that I tried not to break. 
https://github.com/pytorch/pytorch/blob/8d070d23d6685219f3f0bd38885c7142be2ec4fa/torch/_dynamo/utils.py#L469-L503

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+            # subprocs, for example, but it occurs to me that any accounting we'd
+            # attempt in the subprocs is a larger problem.
+            if METRICS_CONTEXT.recording():
+                METRICS_CONTEXT.increment(dynamo_compile_column, time_spent // 1e3)
```
these need unit of measure suffixes

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -219,185 +207,111 @@ def print_time_report() -> None:
+            # CompilationMetrics (but note that we accumulate by the associated event
+            # name, not the field name in CompilationMetrics). Do we want to keep it
+            # this way for now?
+            cumulative_time_spent[event_name] += time_spent
```
these need unit of measure suffixes

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -925,10 +867,7 @@ def record_compilation_metrics(compilation_metrics: CompilationMetrics):
-            k: list(v) if isinstance(v, set) else v
-            for k, v in dataclasses.asdict(compilation_metrics).items()
-        },
+        lambda: {k: list(v) if isinstance(v, set) else v for k, v in metrics.items()},
```
huh what happened here?

masnesral replied:
There's no need to convert the dict to a CompilationMetrics object and then convert it back to a dict with dataclasses.asdict when we already have the dict form. So that's all I changed (but I can also put it back)

ezyang commented on torch/_dynamo/metrics_context.py:
```diff
@@ -0,0 +1,97 @@
+        if self._level == 0:
+            raise RuntimeError("Cannot set metrics outside of a MetricsContext")
+
+    def increment(self, metric: str, value: Union[str, float]) -> None:
```
type annotation here looks almost certainly wrong

masnesral replied:
In Python, you never really know lol

ezyang commented on torch/_dynamo/metrics_context.py:
```diff
@@ -0,0 +1,97 @@
+        if metric in self._metrics:
+            raise RuntimeError(
+                f"Metric '{metric}' has already been set in the current context"
+            )
```
Good, thanks

ezyang replied:
But I think this should be default, and the non-erroring variant banned

masnesral replied:
changed

ezyang commented on torch/_dynamo/metrics_context.py:
```diff
@@ -0,0 +1,97 @@
+        # the get/put timing could be bumped on that helper...
+        self._check()
+        if metric not in self._metrics:
+            self._metrics[metric] = 0
```
You have type confusion here. Is this an integer or a float? You may need dedicated methods for each

masnesral replied:
I think since we standardized on time_ns, we can make it int for now and add a float version later if the need arises.

ezyang reviewed (changes requested):
At a high level this looks great. We just need to polish it more.

ezyang commented:
You need to think about your testing plan, having no testing plan is bad, fairly detailed manual end to end testing is going to be needed.

masnesral commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1991,17 +1993,50 @@ def _backward_impl(ctx, all_args):
+                        "entire_frame_compile",
+                        fn_name="backward._backward_impl",
+                        log_pt2_compile_event=False,
+                        # TODO: How do we want to record this one in dynamo_compile?
```
will make it different.

masnesral committed:
Update on "[RFC] [logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented on test/dynamo/test_utils.py:
```diff
@@ -198,7 +200,7 @@ def test_dynamo_timed(self, mock_time, mock_time_ns):
                    dynamo_time_before_restart_s=0.0,
                    has_guarded_code=True,
-                   remote_cache_time_saved_s=0,
+                   remote_cache_time_saved_s=None,
```
I believe these changes from 0 to None are fine. We're not hitting the remote cache path on the OSS tests so the field is never populated. It's just that the previous code defaulted to 0 when not populated: https://github.com/pytorch/pytorch/blob/e6c5a7748516a2c2f3cffcfb4572603a0578280a/torch/_dynamo/convert_frame.py#L1073

masnesral commented on test/dynamo/test_utils.py:
```diff
@@ -256,7 +261,7 @@ def test_dynamo_timed(self, mock_time, mock_time_ns):
                    has_guarded_code=None,
                    remote_cache_time_saved_s=None,
-                   structured_logging_overhead_s=0.0,
+                   structured_logging_overhead_s=None,
```
Same rationale as above

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -270,188 +261,121 @@ def print_time_report() -> None:
+        if dynamo_compile_column:
+            metrics_context = get_metrics_context()
+            metrics_context.increment(dynamo_compile_column, time_spent // 1000)
+            # TODO: the events that we capture in calculate_time_spent() seem a little
```
If we never want to change this, I can remove the TODO

masnesral committed:
Update on "[RFC] [logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] [logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1449,7 +1448,9 @@ def load_with_key(
-                add_remote_cache_time_saved(time_saved_ns, is_backward)
+                get_metrics_context().increment(
+                    "distributed_ephemeral_timeout_us", time_saved_ns // 1000
+                )
```
note for later: actually `distributed_ephemeral_timeout_us` here is not accurate, the true time saved is `ephemeral_increase` lol

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -139,12 +141,10 @@
-codecache_metrics: Counter[str] = collections.Counter()
+# This supports calculate_time_spent(), which reports cumulative times (in seconds)
+# across the process for any "phase" populated by dynamo_timed. Reset if
+# reset_frame_count() is called.
```
This is seconds because it's the grandfathered second-based APIs, is that right?

masnesral replied:
Ugh, no I named it wrong.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -233,25 +233,16 @@ def increment_op_count(cnt: int) -> None:
-    if total_by_key:
-        total_by_key["total_wall_time"] = total_wall_time
+    for phase, timing in cumulative_time_spent_s.items():
+        total_by_key[phase] = timing / 1e9
```
??? This seems clearly like a unit error. timing seems to be seconds, why you divide by 1e9?

masnesral replied:
bad name

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -233,25 +233,16 @@ def increment_op_count(cnt: int) -> None:
 
+    total_by_key["total_wall_time"] = total_by_key.get(
+        "entire_frame_compile", 0
+    ) + total_by_key.get("entire_backward_compile", 0)
```
This seems kind of suspicious for when autotuning shows up, but maybe this is legacy so we don't care

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -270,188 +261,121 @@ def print_time_report() -> None:
+    if metadata:
+        event_metadata.update(metadata)
+    if fn_name:
+        event_metadata.update({"fn_name": fn_name})
```
OK for bc

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -270,188 +261,121 @@ def print_time_report() -> None:
-                                ),
-                            )
-                            record_compilation_metrics(metrics)
+        time_spent = end_ns - start_ns
```
please `_ns` suffix thanks

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -270,188 +261,121 @@ def print_time_report() -> None:
+        )
+        if dynamo_compile_column:
+            metrics_context = get_metrics_context()
+            metrics_context.increment(dynamo_compile_column, time_spent // 1000)
```
Unit confusion here. It would be good for there to be a `_us` somewhere here making it clear that this is microsecond granularity. Maybe the column name? Or maybe metrics context can have some extra typing. OK to save for later.

masnesral replied:
I can rename the param

ezyang commented on torch/_dynamo/metrics_context.py:
```diff
@@ -0,0 +1,82 @@
+        Increment a metric by a given amount.
+        """
+        if self._level == 0:
+            raise RuntimeError(f"Cannot increment {metric} outside of a MetricsContext")
```
Because of AOTInductor and other "non-standard", currently uninstrumented paths, you have to make this inert when level == 0. Would be OK to follow up and force consistent instrumentation for these paths in the future. This is some of your CI failures.

ezyang replied:
The other case I see in CI is the newly added runtime autotuning, which also is unscoped (it should get scoped, but not in this PR)

ezyang reviewed (approved):
CI failures look easy to fix

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "unrelated ci failure"

kit1980 commented:
revert -m "Breaking an internal tests + there is a bug according to the author" -c ghfirst

kit1980 commented:
@pytorchbot revert -m "Breaking an internal tests + there is a bug according to the author" -c ghfirst

jamesjwu commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -2002,17 +2004,49 @@ def _backward_impl(ctx, all_args):
+                        aot_config, "backward"
+                    ), get_metrics_context(), dynamo_timed(
+                        "backward._backward_impl",
+                        phase_name="entire_backward_compile",
```
Yeah if you can just add that exact line

jamesjwu replied:
```suggestion
                        phase_name="entire_backward_compile",
                        log_pt2_compile_event=True,
```

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Overhaul dynamo_timed and CompilationMetrics logging."


Here's the overview:

There's a new contextmanager singleton called MetricsContext. Entering the MetricsContext is how we demarcate the boundary on which we'll create a single CompilationMetrics object, and therefore, a single dynamo_compile log entry. While we're inside the MetricsContext, we can update/set many different metrics. Most importantly: `dynamo_timed` can also update the in-progress MetricsContext. In the proposal here, we tell `dynamo_timed` that we want it to do so by providing the name of the MetricsContext field to increment. There can be many `dynamo_timed` calls in different parts of the code updating different fields. Then when the MetricsContext exits, that's when the logging of everything gathered finally happens. One potential footgun is trying to use `dynamo_timed` when we haven't entered the MetricsContext, but we assert on that problem. Another problem is that we re-enter the context recursively, but we watch for that and do the logging only when the outermost exits.

Some specifics:
* Introduce MetricsContext - a context manager that on exit, records the CompilationMetrics (which also logs to dynamo_compile).
* Completely remove the concept of frame_phase_timing. Instead, update the MetricsContext during compilation, either directly or via dynamo_timed.
* Remove some globals we previously used to accumulate counters to later populate a CompilationMetrics. We use CompilationMetrics set/update/increment APIs instead.
* `record_compilation_metrics` is now called on exit from MetricsContext.
* Populate legacy CompilationMetrics fields right before logging, inside `record_compilation_metrics`.
* Remove the one-off `add_remote_cache_time_saved` helper; capture that timing directly into the MetricsContext.

And specifically, several changes to dynamo_timed:
* "Modernize" the parameters and update all callsites accordingly.
* Move the backwards logging of the CompilationMetrics to the backwards compile location.
* Add a parameter for which CompilationMetrics field to update

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

ezyang commented:
Surprisingly, this PR improves vision_maskrcnn compile time quite a bit

<img width="348" alt="image" src="https://github.com/user-attachments/assets/a96ed3b8-bef4-47e0-a4df-43cf90ec184e">
